firebase-firestore/src/main/java/com/google/firebase/firestore/model/value/ProtoValues.java

```diff
@@ -249,4 +249,87 @@ private static int compareMaps(MapValue left, MapValue right) {
     // Only equal if both iterators are exhausted.
     return Util.compareBooleans(iterator1.hasNext(), iterator2.hasNext());
   }
+
+  /** Generate the canonical ID for the provided field value (as used in Target serialization). */
+  public static String canonicalId(Value value) {
+    StringBuilder builder = new StringBuilder();
+    buildCanonicalId(builder, value);
+    return builder.toString();
+  }
+
+  // TODO(mrschmidt): Use in target serialization and migrate all existing TargetData
+  private static void buildCanonicalId(StringBuilder builder, Value value) {
+    switch (value.getValueTypeCase()) {
+      case NULL_VALUE:
+        builder.append("null");
+        break;
+      case BOOLEAN_VALUE:
+        builder.append(value.getBooleanValue());
+        break;
+      case INTEGER_VALUE:
+        builder.append(value.getIntegerValue());
+        break;
+      case DOUBLE_VALUE:
+        builder.append(value.getDoubleValue());
+        break;
+      case TIMESTAMP_VALUE:
+        buildTimestampCanonicalId(builder, value.getTimestampValue());
+        break;
+      case STRING_VALUE:
+        builder.append(value.getStringValue());
+        break;
+      case BYTES_VALUE:
+        builder.append(Util.toDebugString(value.getBytesValue()));
+        break;
+      case REFERENCE_VALUE:
+        // TODO(mrschmidt): Use document key only
+        builder.append(value.getReferenceValue());
+        break;
+      case GEO_POINT_VALUE:
+        buildGeoPointCanonicalId(builder, value.getGeoPointValue());
+        break;
+      case ARRAY_VALUE:
+        buildArrayCanonicalId(builder, value.getArrayValue());
+        break;
+      case MAP_VALUE:
+        buildObjectCanonicalId(builder, value.getMapValue());
+        break;
+      default:
+        throw fail("Invalid value type: " + value.getValueTypeCase());
+    }
+  }
+
+  private static void buildTimestampCanonicalId(StringBuilder builder, Timestamp timestamp) {
+    builder.append(String.format("{s:%s,n:%s}", timestamp.getSeconds(), timestamp.getNanos()));
+  }
+
+  private static void buildGeoPointCanonicalId(StringBuilder builder, LatLng latLng) {
+    builder.append(String.format("{lat:%s,lng:%s}", latLng.getLatitude(), latLng.getLongitude()));
+  }
+
+  private static void buildObjectCanonicalId(StringBuilder builder, MapValue mapValue) {
+    builder.append("{");
+    boolean first = true;
+    for (Map.Entry<String, Value> entry : mapValue.getFieldsMap().entrySet()) {
```

Comment:
Are you intending to make it an invariant that MapValues are always sorted by key? I don't recall seeing this in your other PR adding `protovalue.ObjectValue.Builder`.

If not, we probably need to sort the strings here :-(. I think maintaining the invariant is probably better though.

Note that server-supplied MapValues already conform to this invariant.